datafusion/core/src/datasource/file_format/csv.rs

```diff
@@ -324,6 +349,188 @@ fn build_schema_helper(names: Vec<String>, types: &[HashSet<DataType>]) -> Schem
     Schema::new(fields)
 }
 
+impl Default for CsvSerializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Define a struct for serializing CSV records to a stream
+pub struct CsvSerializer {
+    // CSV writer builder
+    builder: WriterBuilder,
+    // Inner buffer for avoiding reallocation
+    buffer: Vec<u8>,
+    // Flag to indicate whether there will be a header
+    header: bool,
+}
+
+impl CsvSerializer {
+    /// Constructor for the CsvSerializer object
+    pub fn new() -> Self {
+        Self {
+            builder: WriterBuilder::new(),
+            header: true,
+            buffer: Vec::with_capacity(4096),
+        }
+    }
+
+    /// Method for setting the CSV writer builder
+    pub fn with_builder(mut self, builder: WriterBuilder) -> Self {
+        self.builder = builder;
+        self
+    }
+
+    /// Method for setting the CSV writer header status
+    pub fn with_header(mut self, header: bool) -> Self {
+        self.header = header;
+        self
+    }
+}
+
+#[async_trait]
+impl BatchSerializer for CsvSerializer {
+    async fn serialize(&mut self, batch: RecordBatch) -> Result<Bytes> {
+        let builder = self.builder.clone();
+        let mut writer = builder.has_headers(self.header).build(&mut self.buffer);
+        writer.write(&batch)?;
+        drop(writer);
+        self.header = false;
+        Ok(Bytes::from(self.buffer.drain(..).collect::<Vec<u8>>()))
+    }
+}
+
+/// This macro tries to abort all writers before returning a possible error.
+macro_rules! handle_err_or_continue {
+    ($result:expr, $writers:expr) => {
+        match $result {
+            Ok(value) => Ok(value),
+            Err(e) => {
+                // Abort all writers before returning the error:
+                for writer in $writers {
+                    let abort_future = writer.abort_writer()?;
```

Comment:
I think this could be a function (which would make it easier to work with). Maybe like

```rust
fn check_for_errors(result: Result<RecordBatch>) -> Result<RecordBatch> {
...
}
```

Also, reading it I am not sure about the error handling. Specifically, 
1.  if the abort fails, the abort error may be reported rather then the initial write error.
2. If canceling one writer fails, the others may not be aborted (as `?` returns early) 

I have some ideas on how to refactor this (into a struct) and to make it more testable,  which I will try and put up as a draft PR shortly

Reply:
I converted macro, to function. Also during abortion we receive an error, that case doesn't return early. We try to abort all writers